src/uopool/services/simulation.rs

```diff
@@ -86,46 +91,220 @@ where
+                        entity: LEVEL_TO_ENTITY[index].to_string(),
+                        opcode: CREATE2_OPCODE.to_string(),
                     });
                 }
             }
+        }
+
+        Ok(())
+    }
 
-            if let Some(count) = trace.number_levels[index].opcodes.get(&*CREATE2_OPCODE) {
-                if LEVEL_TO_ENTITY[&index] == "factory" && *count == 1 {
+    fn parse_slots(
+        &self,
+        keccak: Vec<Bytes>,
+        stake_info_by_entity: &HashMap<usize, StakeInfo>,
+        slots_by_entity: &mut HashMap<Address, HashSet<String>>,
+    ) {
+        for kecc in keccak {
+            for entity in stake_info_by_entity.values() {
+                if entity.address.is_zero() {
                     continue;
                 }
-                return Err(SimulateValidationError::OpcodeValidation {
-                    entity: LEVEL_TO_ENTITY[&index].to_string(),
-                    opcode: CREATE2_OPCODE.to_string(),
-                });
+
+                let entity_address_bytes =
+                    Bytes::from([vec![0; 12], entity.address.to_fixed_bytes().to_vec()].concat());
+
+                if kecc.starts_with(&entity_address_bytes) {
+                    let k = AbiEncode::encode_hex(keccak256(kecc.clone()));
+                    slots_by_entity
+                        .entry(entity.address)
+                        .or_insert(HashSet::new())
+                        .insert(k);
+                }
+            }
+        }
+    }
+
+    fn associated_with_slot(
+        &self,
+        address: &Address,
+        slot: &String,
+        slots_by_entity: &HashMap<Address, HashSet<String>>,
+    ) -> Result<bool, SimulateValidationError<M>> {
+        if *slot == address.to_string() {
+            return Ok(true);
+        }
+
+        if !slots_by_entity.contains_key(address) {
+            return Ok(false);
+        }
+
+        let slot_as_number = U256::from_str_radix(slot, 16)
+            .map_err(|_| SimulateValidationError::StorageAccessValidation { slot: slot.clone() })?;
+
+        if let Some(slots) = slots_by_entity.get(address) {
+            for slot_entity in slots {
+                let slot_entity_as_number =
+                    U256::from_str_radix(slot_entity, 16).map_err(|_| {
+                        SimulateValidationError::StorageAccessValidation { slot: slot.clone() }
+                    })?;
+
+                if slot_as_number >= slot_entity_as_number
+                    && slot_as_number < (slot_entity_as_number + 128)
+                {
+                    return Ok(true);
+                }
+            }
+        }
+
+        Ok(false)
+    }
+
+    fn storage_access(
+        &self,
+        user_operation: &UserOperation,
+        entry_point: &Address,
+        stake_info_by_entity: &HashMap<usize, StakeInfo>,
+        trace: &JsTracerFrame,
+    ) -> Result<(), SimulateValidationError<M>> {
+        let mut slots_by_entity = HashMap::new();
```

Comment:
Is it possible to use just `HashMap<Address, HashSet<Bytes>>` instead of `HashMap<Address, HashSet<String>>` which we could skip some bytes to string and string to bytes converts? I am just curious.

Reply:
Nice catch, that makes more sense.